src/jdk.jdwp.agent/share/native/libjdwp/util.c

```diff
@@ -1099,31 +1106,394 @@ debugMonitorNotifyAll(jrawMonitorID monitor)
+ * we iterate over all monitors, and need to make sure the monitor we are checking has
+ * been initialized. If it hasn't been, then it is skipped. There can be a race and
+ * we catch it in the middle of intialization. Since dbgRawMonitor is held during
+ * verifyMonitorRank(), we should also hold it during monitor initialization.
+ */
+
+static jrawMonitorID dbgRawMonitor;
+
+static void
+createDbgRawMonitor()
+{
+    JDI_ASSERT(gdata->rankedMonitors);
+    jvmtiError error;
+    error = JVMTI_FUNC_PTR(gdata->jvmti,CreateRawMonitor)
+            (gdata->jvmti, "Debug Raw Monitor", &dbgRawMonitor);
+    if (error != JVMTI_ERROR_NONE) {
+        EXIT_ERROR(error, "on creation of dbgRawMonitor");
+    }
+    JDI_ASSERT(dbgRawMonitor != NULL);
+}
+
+void
+dbgRawMonitor_lock()
+{
+    JDI_ASSERT(gdata->rankedMonitors);
+    jvmtiError error;
+    error = JVMTI_FUNC_PTR(gdata->jvmti,RawMonitorEnter)
+            (gdata->jvmti, dbgRawMonitor);
+    error = ignore_vm_death(error);
+    if (error != JVMTI_ERROR_NONE) {
+      EXIT_ERROR(error, "on dbgRawMonitor enter");
+    }
+}
+
+void
+dbgRawMonitor_unlock()
+{
+    JDI_ASSERT(gdata->rankedMonitors);
+    jvmtiError error;
+    error = JVMTI_FUNC_PTR(gdata->jvmti,RawMonitorExit)
+            (gdata->jvmti, dbgRawMonitor);
+    error = ignore_vm_death(error);
+    if (error != JVMTI_ERROR_NONE) {
+      EXIT_ERROR(error, "on dbgRawMonitor exit");
+    }
+}
+
+DebugRawMonitor*
+debugMonitorCreate(DebugRawMonitorRank dbg_monitor_rank, char *name)
 {
     jrawMonitorID monitor;
     jvmtiError error;
+    DebugRawMonitor* dbg_monitor = &dbg_monitors[dbg_monitor_rank];
+    JDI_ASSERT(dbg_monitor->monitor == NULL);
 
     error = JVMTI_FUNC_PTR(gdata->jvmti,CreateRawMonitor)
                 (gdata->jvmti, name, &monitor);
     if (error != JVMTI_ERROR_NONE) {
         EXIT_ERROR(error, "on creation of a raw monitor");
     }
-    return monitor;
+
+    JDI_ASSERT(monitor != NULL);
+
+    // Need to lock during initialization so verifyMonitorRank() can be guaranteed that
+    // if the monitor field is set, then the monitor if fully initialized. More specifically,
+    // that the rank field has been set.
+    if (gdata->rankedMonitors) {
+        dbgRawMonitor_lock();
+    }
+    dbg_monitor->monitor = monitor;
+    dbg_monitor->rank = dbg_monitor_rank;
+    dbg_monitor->name = name;
+    if (gdata->rankedMonitors) {
+        dbgRawMonitor_unlock();
+    }
+
+    return dbg_monitor;
 }
 
 void
-debugMonitorDestroy(jrawMonitorID monitor)
+debugMonitorDestroy(DebugRawMonitor *dbg_monitor)
 {
     jvmtiError error;
+    JDI_ASSERT(dbg_monitor != NULL);
 
     error = JVMTI_FUNC_PTR(gdata->jvmti,DestroyRawMonitor)
-                (gdata->jvmti, monitor);
+                (gdata->jvmti, dbg_monitor->monitor);
     error = ignore_vm_death(error);
     if (error != JVMTI_ERROR_NONE) {
         EXIT_ERROR(error, "on destruction of raw monitor");
     }
+
+    dbg_monitor->monitor = NULL;
```

Comment:
I don't see how that helps. Access to the field is not protected.

Reply:
I guess there could be a race if one thread is destroying this monitor while another is trying to use it. Thus a thread could be doing something like a RawMonitorEnter in the middle of (or after) the monitor being destroyed. Fixing that would require holding dbgRawMonitor during RawMonitorEnter, and that would cause deadlocks all over the place. Also, this same potential issue exists already, but doesn't seem to ever arise. It seems we only call debugMonitorDestroy for cmdQueueLock. Not sure why that is.

Reply:
FWIW Deleting monitors is a tricky business and needs to be done with great care. You have to ensure all threads using the monitor are completely done with it. Simply locking it first to check it is "unused" is not sufficient.

Reply:
This is probably the reason that the debug agent only destroys one of the 20 or so monitors it creates. I'm not sure how it got to that point. It may have destroyed more at one point but there were issues, and only one survived destruction without issue.

Reply:
If debugger attaches to the debuggee, detaches and re-attaches again, are the monitors recreated again?
(with rankedMonitor you added an assert if `DebugRawMonitor::monitor` is not null)

Reply:
> If debugger attaches to the debuggee, detaches and re-attaches again, are the monitors recreated again?
> (with rankedMonitor you added an assert if `DebugRawMonitor::monitor` is not null)

Almost all of the debugMonitorCreate() calls seem to come (indirectly) from initialize(), which is only called once.

cmdQueueLock, which is the one monitor we call debugMonitorDestroy() on, is called from debugLoop_run(), which is called whenever a debugger connection is initiated. This is the only case of creating, destroying, and then recreating. I wonder if there is some need for that.

popFrameEventLock and popFrameProceedLock are created lazily when first needed. There is a check to make sure they are not recreated if already created. They are never destroyed.

That seems to be it. So we pretty much create all monitors upon initialization and never recreate those that are created during initialization.

With regard to the assert, this can only be an issue for cmdQueueLock, and since  debugMonitorDestroy() is called on it (and that clears the needed fields), the assert shouldn't be triggered.